asmcomp/arm64/emit.mlp

```diff
@@ -767,34 +764,40 @@ let emit_instr env i =
               reg_tmp1 in
         begin match memory_chunk with
         | Byte_unsigned ->
-            `	{emit_string load_type}b	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrb	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Byte_signed ->
-            `	{emit_string load_type}sb	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrsb	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Sixteen_unsigned ->
-            `	{emit_string load_type}h	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrh	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Sixteen_signed ->
-            `	{emit_string load_type}sh	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrsh	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Thirtytwo_unsigned ->
-            `	{emit_string load_type}	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldr	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Thirtytwo_signed ->
-            `	{emit_string load_type}sw	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrsw	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Single ->
-            `	{emit_string load_type}	s7, {emit_addressing addressing_mode base}\n`;
+            `	ldr	s7, {emit_addressing addressing_mode base}\n`;
             `	fcvt	{emit_reg dst}, s7\n`
-        | Word_int | Word_val | Double ->
-            `	{emit_string load_type}	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+        | Word_int | Word_val ->
+            if is_atomic then begin
+              assert (addressing_mode = Iindexed 0);
+              `	ldar	{emit_reg dst}, [{emit_reg i.arg.(0)}]\n`
```

Comment:
So I was double checking the compilation of this:
```ocaml
let code0 x y = Atomic.set x 1 ; Atomic.get y
```
and I see:
```
_camlSBA__code0_787:
        sub     sp, sp, #0x10
        str     x30, [sp, #0x8]
        mov     x20, x1
        orr     x1, xzr, #0x3
        mov     x19, sp
        ldr     x16, [x28, #0x30]
        mov     sp, x16
        bl      _caml_atomic_exchange
        mov     sp, x19
        ldar    x0, [x20]
        ldr     x30, [sp, #0x8]
        add     sp, sp, #0x10
        ret
```
Shouldn't the atomic load give us:
```
        dmb     ishld
        ldar    x0, [x20]
```
as per #10995. (and I think clarified here: https://github.com/ocaml/ocaml/pull/10972#issuecomment-1032940299)

I think the change that is needed is:
```suggestion
              ` dmb ishld\n`
              `	ldar	{emit_reg dst}, [{emit_reg i.arg.(0)}]\n`
```

Reply:
Yes, sure.  I thought I pushed that change already?  Apparently not :-(

Reply:
PS.  When you're happy with this PR, I need to clean up the history!

Reply:
ouch, apologies, merged it up. 
Is it more than squashing?

Reply:
Just squashing.  We can do it when merging https://github.com/ocaml/ocaml/pull/10972